functions/src/aqi-calculation/index.ts

```diff
@@ -52,36 +59,67 @@ exports.thingspeakToFirestore = functions
         thingspeakInfo
       );
 
-      const resolvedPath = readingsSubcollection(knownSensor.id);
-
-      // Only add data if not already present in database.
-      // This happens if a sensor is down, so only old data is returned.
+      const docId = knownSensor.id;
+      const resolvedPath = readingsSubcollection(docId);
       const readingsRef = firestore.collection(resolvedPath);
+
+      // If data is not already present in the database add
+      // it to the historical readings
+      let firestoreSafeReading = defaultPm25BufferElement;
       if (
         (
           await readingsRef
             .where('timestamp', '==', Timestamp.fromDate(reading.timestamp))
             .get()
         ).empty
       ) {
-        const firestoreSafeReading = {
+        // Update reading with values
+        firestoreSafeReading = {
           timestamp: Timestamp.fromDate(reading.timestamp),
           channelAPm25: reading.channelAPm25,
           channelBPm25: reading.channelBPm25,
           humidity: reading.humidity,
           latitude: reading.latitude,
           longitude: reading.longitude,
         };
+        // Add to historical readings
         await readingsRef.add(firestoreSafeReading);
       }
 
-      // Delays the loop so that we hopefully don't overload Thingspeak, avoiding
-      // our program from getting blocked.
-      // Allocates up to a minute of the two minute runtime for delaying
-      const oneMinuteInMilliseconds = 60000;
-      const delayBetweenSensors = oneMinuteInMilliseconds / sensorList.length;
-      await new Promise(resolve => setTimeout(resolve, delayBetweenSensors));
+      // Add readings to the PM 2.5 buffer
+      const sensorDocRef = firestore.collection('sensors').doc(knownSensor.id);
+      const status =
+        knownSensor.data().pm25BufferStatus ?? bufferStatus.DoesNotExist;
```

Comment:
Since we call knownSensor.data() so much here, let's create a separate variable for it